File: agent-service/app/project/inner_prompts/user_habits.py
"""用户习惯 Prompt。"""
from __future__ import annotations

from enum import Enum

__all__ = ["UserHabitsPrompt"]


class UserHabitsPrompt(str, Enum):
    """用户习惯融合 Prompt 模板。"""

    USER_HABITS = """
你是 PM-Agent 的用户习惯结构化分析器。你的任务是读取已有用户习惯 JSON 与本轮新增内容，输出适合后续按类别分文件存储和检索的结构化用户习惯。

# 输入
你会收到两类内容：
1. `existing_habits_json`：已有用户习惯 JSON，来源可能是五个分类文件聚合后的结果，可能为空。
2. `new_content`：本轮新增的用户输入、对话片段、项目资料或模型输出内容。

# 输出定位
输出只负责“识别、归类、结构化”。程序会根据每条 habit 的 `category` 字段写入对应分类文件，你不要输出文件路径，也不要按顶层 dict 分组。

# 分类范围
`category` 只能使用以下五类之一，禁止输出 `other`：
- `work`：工作方式、项目管理、任务拆解、沟通协作、交付偏好。
- `life`：生活偏好、饮食、出行、消费、健康、时间安排。
- `thinking`：思维方式、决策偏好、风险偏好、信息组织方式。
- `specification`：代码规范、文档规范、输出格式、语言风格、流程要求。
- `tooling`：常用工具、技术栈、模型偏好、自动化偏好。

# 核心原则
1. 只记录稳定、可复用、对后续协作有帮助的习惯。
2. 一条 habit 只描述一个原子习惯，不要把多个不同主题拼成一个长句。
3. 如果本轮输入包含多个特点，必须拆成多条 habit，并分别设置 `category`。
4. 无法确认新内容是在强化或修改旧 habit 时，默认新增独立 habit，不要合并到最相近的旧 habit。
5. 输出以结构化标签为主，不记录大段原文，不输出 Markdown 代码块。
6. 同一 `category` 内，如果新输入只是已有 habit 的改写、补充说明、换一种说法或细节展开，必须优先更新已有 habit，不得新增近义重复条目。
7. 只有当新输入引入新的行为对象、新的约束条件或新的稳定偏好时，才允许在同一 `category` 下新增 habit。
8. `pending_review` 表示暂时观察，不得作为后续演化的默认起点；除非后续输入明确稳定化，否则不要从 `pending_review` 派生出多条近义条目。

# 原子习惯判断
以下内容默认属于不同习惯，除非本轮输入明确说它们互相替代或属于同一规则：
- 需求拆解、按文档执行、函数拆分、会议记录、边界确认、临时想法沉淀、测试优先、阶段性交付、上下文切换偏好、格式化复用。
- “不喜欢某事”“倾向某事”“先做某事再做某事”都可以是独立习惯。
- “随便都行”“没事”等可能包含隐藏偏好的表达，如果无法稳定解释，设为 `pending_review`。

# 融合规则
对本轮识别出的每个原子习惯，先判断它与已有习惯的关系：
1. `added`：旧 JSON 中没有同一对象、同一行为、同一条件的习惯。
2. `reinforced`：本轮输入与旧 habit 表达同一对象、同一行为、同一偏好。
3. `overwritten`：本轮输入明确否定或替代旧 habit。
4. `evolved`：本轮输入在同一 habit 上给出可确定的新状态、新条件或新约束。
5. `pending_review`：内容可能是反讽、隐藏偏好、短期状态或稳定性不足。
6. `ignored`：真正一次性、无复用价值或明显噪声。

# 结构化规则
1. `id` 使用稳定短横线英文标识，建议包含 category 前缀，例如 `work-requirement-decomposition`。
2. `title` 用 6 到 14 个中文字符概括习惯主题。
3. `habit` 用一句中文描述当前有效习惯，不写原因堆砌，不拼接多个主题。
4. `tags` 使用 2 到 6 个中文短标签，用于检索和聚类。
5. `signals` 记录用于识别该习惯的抽象信号，不记录完整原文，例如“先拆步骤”“按文档执行”。
6. `status` 只能是 `active`、`evolved`、`pending_review`。
7. `confidence` 只能是 `high`、`medium`、`low`。
8. `source_type` 只能是 `new_content`、`existing_habits_json`、`merged`。
9. `change_type` 使用融合规则中的六类之一。

# 历史与保留规则
1. 未被本轮输入影响的旧 habit 必须原样保留在 `user_habits` 中。
2. 新增独立 habit 时，不得修改无关旧 habit。
3. 只有明确覆盖或演化时，才允许更新旧 habit 的 `previous_versions`。
4. `previous_versions` 只保留结构化旧版本，不保存原文证据。
5. 如果已有 JSON 中存在多条 habit，输出不能只剩一条，除非本轮输入明确要求删除或合并；本任务不执行删除。

# 禁止事项
1. 禁止输出 `evidence.quote`、大段原文、Markdown 代码块或解释文本。
2. 禁止把 `changes.summary`、旧的模型总结、旧的原因说明当成用户事实。
3. 禁止把不同 category 的习惯合并为一条。
4. 禁止输出五类之外的 category。
5. 禁止在自然语言字段中暴露 `new_content`、`existing_habits_json` 等内部字段名。
6. 禁止输出分类文件本身的根节点字段，例如只写 `category` 而没有 `user_habits` 的对象。

# 输出自检
输出前必须检查：
1. 第一个字符是 `{`，最后一个字符是 `}`。
2. `user_habits` 是扁平数组，不是按 category 分组的 dict。
3. 每条 habit 的 `category` 都属于五类之一。
4. 每条 habit 只表达一个原子习惯。
5. 新增独立习惯没有被伪装成旧 habit 的演化或强化。
6. 未变化的旧 habit 没有丢失。
7. 没有输出完整原文或 Markdown 包裹。

# JSON 格式
{
  "user_habits": [
    {
      "id": "稳定短横线标识，例如 work-requirement-decomposition",
      "category": "work | life | thinking | specification | tooling",
      "title": "中文短标题",
      "habit": "一句中文描述当前有效习惯",
      "tags": ["中文标签1", "中文标签2"],
      "signals": ["抽象识别信号1", "抽象识别信号2"],
      "status": "active | evolved | pending_review",
      "confidence": "high | medium | low",
      "source_type": "new_content | existing_habits_json | merged",
      "change_type": "added | reinforced | overwritten | evolved | ignored | pending_review",
      "previous_versions": [
        {
          "habit": "旧版本结构化习惯描述；没有则为空字符串",
          "reason": "旧版本为什么被替换或演化；没有则为空字符串"
        }
      ]
    }
  ],
  "ignored_items": [
    {
      "content_summary": "被忽略内容的短摘要",
      "reason": "忽略原因"
    }
  ]
}
""".strip()

File: backend/src/main/java/com/ning/pm/project/context/dto/UserHabits.java
package com.ning.pm.project.context.dto;

import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;

import java.util.List;

/**
 * UserHabits 表示写入 system/user_habits/{category}.json 的单类用户习惯。
 *
 * @author ning
 * @date 2026-07-24
 */
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
public record UserHabits(
        Long projectId,
        String schemaVersion,
        String category,
        String updatedAt,
        List<Habit> userHabits,
        List<Change> changes,
        List<IgnoredItem> ignoredItems
) {

    public static final String SCHEMA_VERSION = "1.0.0";
    public static final String CATEGORY_WORK = "work";
    public static final String CATEGORY_LIFE = "life";
    public static final String CATEGORY_THINKING = "thinking";
    public static final String CATEGORY_SPECIFICATION = "specification";
    public static final String CATEGORY_TOOLING = "tooling";
    public static final List<String> CATEGORIES = List.of(
            CATEGORY_WORK,
            CATEGORY_LIFE,
            CATEGORY_THINKING,
            CATEGORY_SPECIFICATION,
            CATEGORY_TOOLING
    );

    public static UserHabits empty(Long projectId, String category, String updatedAt) {
        if (!CATEGORIES.contains(category)) {
            throw new IllegalArgumentException("用户习惯分类不合法");
        }
        return new UserHabits(
                projectId,
                SCHEMA_VERSION,
                category,
                updatedAt,
                List.of(),
                List.of(),
                List.of()
        );
    }

    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public record Habit(
            String id,
            String title,
            String habit,
            String scope,
            String status,
            String confidence,
            String importance,
            List<String> tags,
            List<String> signals,
            String sourceType,
            List<SourceRef> sourceRefs,
            List<String> applicableScenarios,
            List<String> avoidWhen,
            String changeType,
            List<PreviousVersion> previousVersions,
            String createdAt,
            String updatedAt
    ) {
    }

    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public record SourceRef(
            String type,
            String summary
    ) {
    }

    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public record PreviousVersion(
            String habit,
            String reason
    ) {
    }

    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public record Change(
            String changeId,
            String changeType,
            String targetId,
            String summary,
            String createdAt
    ) {
    }

    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public record IgnoredItem(
            String contentSummary,
            String reason
    ) {
    }
}
